packages/tools/guiEditor/src/diagram/gizmoLine.tsx

```diff
@@ -32,45 +81,224 @@ export function GizmoLine(props: IGizmoLineProps) {
+
+        const rotatedDelta = new Vector2();
+        delta.rotateToRef(control.rotation, rotatedDelta);
+
+        const deltaX = rotatedDelta.x;
+        const deltaY = rotatedDelta.y;
+
+        // Move only the scale point that was touched
+        const movedPointIndex = movedScalePoint.current;
+
+        if (movedPointIndex === 0) {
+            // Moved first point, (x1, y1)
+            let x1 = control._x1.getValue(control._host);
+            x1 += deltaX;
+            control.x1 = MathTools.Round(x1);
+
+            let y1 = control._y1.getValue(control._host);
+            y1 += deltaY;
+            control.y1 = MathTools.Round(y1);
+        } else if (movedPointIndex === 2) {
+            let x2 = control._x2.getValue(control._host);
+            x2 += deltaX;
+            control.x2 = MathTools.Round(x2);
+
+            let y2 = control._y2.getValue(control._host);
+            y2 += deltaY;
+            control.y2 = MathTools.Round(y2);
+        }
+        globalState.onPropertyGridUpdateRequiredObservable.notifyObservers();
+    };
+
+    const _rotateEndpoint = (currentPointer: Vector2) => {
+        const currentToPivot = currentPointer.subtract(pivot.current);
+        currentToPivot.normalize();
+        const lastToPivot = lastCursor.current.subtract(pivot.current);
+        lastToPivot.normalize();
+
+        const dotProd = Vector2.Dot(currentToPivot, lastToPivot);
+        const angle = Math.acos(dotProd);
+
+        const direction = -Math.sign(currentToPivot.x * lastToPivot.y - currentToPivot.y * lastToPivot.x);
+
+        if (!isNaN(angle)) {
+            control.rotation += direction * angle;
+            control.rotation = MathTools.Round(control.rotation);
+            globalState.onPropertyGridUpdateRequiredObservable.notifyObservers();
+        }
+    };
+
+    const onMove = () => {
+        const scene = globalState.workbench._scene;
+        const currentPointer = new Vector2(scene.pointerX, scene.pointerY);
+        if (isDragging.current && isPivotMoving.current) {
+            _dragPivot(currentPointer);
+        } else if (isDragging.current) {
+            _dragEndpoint(currentPointer);
+        } else if (isRotating.current) {
+            _rotateEndpoint(currentPointer);
+        }
+        lastCursor.current = currentPointer;
+    };
+
+    const onUp = () => {
+        isDragging.current = false;
+        isRotating.current = false;
+    };
+
+    React.useEffect(() => {
+        const pointerMoveObserver = props.globalState.onPointerMoveObservable.add(onMove);
+        const pointerUpObserver = props.globalState.onPointerUpObservable.add(onUp);
+        return () => {
+            props.globalState.onPointerMoveObservable.remove(pointerMoveObserver);
+            props.globalState.onPointerUpObservable.remove(pointerUpObserver);
+        };
+    }, []);
+
+    const onDrag = (event?: React.PointerEvent<HTMLDivElement>, scalePoint?: IScalePoint) => {

```

Comment:
This function is probably being called very often, but it created many objects on each call. I know this is a final product and not the core framework itself, but we should stick to our best practices and reuse as much as possible, where possible.